### PR TITLE
feat: add job run command

### PR DESCRIPTION
**Pull request by kangere**
This PR adds a `job run` command which can be used to execute scheduled jobs

Addresses #3469 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

**Comment by kangere** [comment]
I had to make some manual changes to resolve the merge conflicts (I haven't worked on this for a while)

File: internal/pkg/aws/stepfunctions/stepfunctions_test.go
Comment (Lou1415926) [inline internal/pkg/aws/stepfunctions/stepfunctions_test.go:73]: Thanks for the tests!!

File: internal/pkg/aws/stepfunctions/stepfunctions.go
Comment (Lou1415926) [inline internal/pkg/aws/stepfunctions/stepfunctions.go]: Do you think this may make the error message more helpful:
```suggestion
		return fmt.Errorf("execute state machine %s: %w", stateMachineARN, err)
```

File: internal/pkg/cli/interfaces.go
Comment (Lou1415926) [inline internal/pkg/cli/interfaces.go]: Seems like we can keep this interface private:
```suggestion
type runner interface {
```

File: internal/pkg/runner/job/job_runner.go
Comment (Lou1415926) [inline internal/pkg/runner/job/job_runner.go]: Same here and the other interfaces in this package! We can keep this unexported if it's not referenced in other packages:
```suggestion
type jobExecutor interface {
```

File: internal/pkg/runner/job/job_runner.go
Comment (Lou1415926) [inline internal/pkg/runner/job/job_runner.go]: For stack name, perhaps we can use this function https://github.com/aws/copilot-cli/blob/efe855432195e92927263dc2643a2d6e35ed8c9c/internal/pkg/deploy/cloudformation/stack/name.go#L28  so that it reuses the code and deal with potentially truncated stack name.

File: internal/pkg/runner/job/job_runner.go
Comment (Lou1415926) [inline internal/pkg/runner/job/job_runner.go]: Similar to the other comment about including info error messages to make them more helpful:
```suggestion
		return fmt.Errorf("describe stack %s: %v", NameForService(r.App, r.Env, r.Job), err)
```

File: internal/pkg/runner/job/job_runner.go
Comment (Lou1415926) [inline internal/pkg/runner/job/job_runner.go]: We can use `aws.StringValue(resource.ResourceType)` which deals with nil-pointers automatically:
```suggestion
		if aws.StringValue(resource.ResourceType) == "AWS::StepFunctions::StateMachine" {
```

Comment (Lou1415926) [reply internal/pkg/runner/job/job_runner.go]: In addition, what do you think of defining a const
```
const CFNStackResourceTypeStateMachine = "AWS::StepFunctions::StateMachine"
```
and use it here?

Perhaps the constant can be defined around here: https://github.com/aws/copilot-cli/blob/efe855432195e92927263dc2643a2d6e35ed8c9c/internal/pkg/deploy/cloudformation/cloudformation.go#L39

File: internal/pkg/runner/job/job_runner.go
Comment (Lou1415926) [inline internal/pkg/runner/job/job_runner.go]: Samesies! For peace of mind we can do:

```suggestion
			stateMachineARN = aws.StringValue(resource.PhysicalReourceId)
```

**Review by Lou1415926 — COMMENTED** [review]
Yayy this is amazing - thanks a ton for writing up the tests and of course the contribution!! I have only some minor comments.

File: internal/pkg/runner/job/job_runner.go
Comment (kangere) [reply internal/pkg/runner/job/job_runner.go]: This function takes in a service name as one of its parameters, would it still work if you pass in the job name?

Comment (Lou1415926) [reply internal/pkg/runner/job/job_runner.go]: Yes! The function name is a bit outdated as it really should be `NameForWorkload` 😅 

Service and job's names are both determined using this function: https://github.com/aws/copilot-cli/blob/966c97f0f0824dace2ab0ebbd76441aa664b6e20/internal/pkg/deploy/cloudformation/stack/workload.go#L131

Comment (kangere) [reply internal/pkg/runner/job/job_runner.go]: okay, my latest commit should address this

File: internal/pkg/aws/stepfunctions/stepfunctions.go
Comment (efekarakus) [inline internal/pkg/aws/stepfunctions/stepfunctions.go]: nit:
```suggestion
//Execute starts a state machine execution.
func (s *StepFunctions) Execute(arn string) error {
```

File: internal/pkg/cli/interfaces.go
Comment (efekarakus) [inline internal/pkg/cli/interfaces.go]: Is it possible to rename the existing `runner` interface to `execRunner`, and `simpleRunner` to `runner` please? 🥺 🙏

File: internal/pkg/cli/job_run.go
Comment (efekarakus) [inline internal/pkg/cli/job_run.go]: 1. For running the job, we should use the credentials from the `EnvManagerRole` instead of the default session. This is because the `default` session might point to an AWS account different than the environment's account where the step function state machine is located.

Can we do something similar to the following to fetch the session from the env manager role:
https://github.com/aws/copilot-cli/blob/a90c9a56909c0addf3e22d1aa9c3c48358799d45/internal/pkg/cli/svc_exec.go#L186-L192

Comment (efekarakus) [reply internal/pkg/cli/job_run.go]: 2. Instead of passing in the `Sess` to the `jobRunner` is it possible to instead use dependency injection?

```go
opts.initRunner = func() error {
   sess, err := o.envSession()
   if err != nil {
      return err
   }
   cfn := cloudformation.New(sess)
   sfn := stepfunctions.New(sess)
   opts.runner =  jobrunner.NewJobRunner({
       Executor: sfn,
       StackRetriever: cfn,
       ...
   })
   return nil
}
```

File: internal/pkg/cli/job_run.go
Comment (efekarakus) [inline internal/pkg/cli/job_run.go]: Can we select the `Job` and `Environment` from a `ConfigSelector` instead of the workspace?
We should be able to run `job run` even if the job manifest file isn't available locally.

```
sel: selector.NewConfigSelector(prompt.New(), configStore),
```

File: internal/pkg/cli/job_run.go
Comment (efekarakus) [inline internal/pkg/cli/job_run.go]: nit: can we rename the package to be `runner/jobrunner` instead?
```suggestion
	"github.com/aws/copilot-cli/internal/pkg/runner/jobrunner"
```

File: internal/pkg/runner/job/job_runner.go
Comment (efekarakus) [inline internal/pkg/runner/job/job_runner.go]: nit: https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences
> Comments should begin with the name of the thing being described and end in a period:
```suggestion
// Package jobrunner provides support for invoking scheduled jobs.
```

File: internal/pkg/runner/job/job_runner.go
Comment (efekarakus) [inline internal/pkg/runner/job/job_runner.go]: Can we add doc comments to all exported names please 🥺 
https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

File: internal/pkg/runner/job/job_runner.go
Comment (efekarakus) [inline internal/pkg/runner/job/job_runner.go]: all of the fields in the `JobRunner` can be private I think :)

File: internal/pkg/runner/job/job_runner.go
Comment (efekarakus) [inline internal/pkg/runner/job/job_runner.go]: nit:
```suggestion
func New(cfg *Config) *JobRunner {
```

**Review by efekarakus — COMMENTED** [review]
Thank you so much @kangere ! this is amazing work 😍 

Please let me know if you need any help with the feedback, we'd be happy to jump in and add additional commits to the PR

File: internal/pkg/cli/interfaces.go
Comment (kangere) [reply internal/pkg/cli/interfaces.go]: sure, that makes sense

File: internal/pkg/cli/job_run.go
Comment (kangere) [reply internal/pkg/cli/job_run.go]: I am not opposed to that idea as I am not as familiar with how sessions work but in the svc_exec.go file you linked the default session is used for initialization purposes, should this have been done differently here ?
https://github.com/aws/copilot-cli/blob/a90c9a56909c0addf3e22d1aa9c3c48358799d45/internal/pkg/cli/svc_exec.go#L63-L70

File: internal/pkg/runner/job/job_runner.go
Comment (kangere) [reply internal/pkg/runner/job/job_runner.go]: I added comments, but not too sure of the wording, feel free to edit them

File: internal/pkg/cli/job_run.go
Comment (kangere) [reply internal/pkg/cli/job_run.go]: This should be addressed in my latest commit

File: internal/pkg/cli/job_run.go
Comment (efekarakus) [reply internal/pkg/cli/job_run.go]: Hi @kangere !

I've addressed all the feedback, there is one more small action that needs to do before we merge the change.
I've added a TODO in `job_run.go` which I'll handle tomorrow morning.

Thank you so much for the contribution! There shouldn't be any more action items from your end :)

Comment (kangere) [reply internal/pkg/cli/job_run.go]: Hi @efekarakus 
Thanks for all the feedback and help. I appreciate it.

I hope I was able to address the comments and feedback well enough. 

I also hope to make many more future contributions!

Comment (efekarakus) [reply internal/pkg/cli/job_run.go]: > I also hope to make many more future contributions!

We would love to, thank you @kangere !

The PR is at a reviewable state, I'll wait for feedback from my teammates and then we can merge it

File: internal/pkg/cli/job_deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/job_deploy.go]:
```suggestion
```

because of #3698 !

**Review by Lou1415926 — APPROVED** [review]
LGTM thank you so much!

**Review by efekarakus — APPROVED** [review]

**Pull request merged**
